[WIP] ansible-test: allow version_added tests for collections

felixfontein opened the pull request:
##### SUMMARY
IMO this is necessary for proper module documentation.

This is a WIP since it builds on top of #66920 for some collection-related data handling. I've tested it manually with `community.crypto`, and it works as expected.

There is now a new option, `--collection-version_added`, which uses the collection's version by default (extracted from `galaxy.yml` resp. `MANIFEST.json`), or an explicit value like `1.0.0` when the form `--collection-version_added 1.0.0` is used. This allows collections to chose their own strategy on what kind of values `version_added` should get.

For this to work, both `--base-branch` and `--docker-keep-git` have to be specified (the latter only when running with `--docker`), like this:
```
ansible-test sanity --docker --docker-keep-git -v --collection-version_added 1.0.0 --base-branch master --test validate-modules
```

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-test

felixfontein committed:
Allow to enable version_added checks for collections.
Always validate version_added on module level if it is there.
Validate all version numbers via schema.

s-hertel commented on test/lib/ansible_test/_data/collection_metadata.py:
```diff
@@ -0,0 +1,57 @@
+    """
+    result = dict(warnings=[])
+
+    # Try reading galaxy.yml
```
If both exist, should the MANIFEST.json take precedence over a galaxy.yml file instead? The manifest indicates a built/installed collection artifact so `ansible-galaxy` is going to prefer the artifact data and fall back to galaxy.yml.

felixfontein replied:
Good idea. That's in the code from #66920 though, so I'll fix it there first and then rebase.

felixfontein replied:
In this PR, it's now in 71c4471b44c4e0bdf65ea2bdf1f5bf6e2329979c.

felixfontein commented:
/rebuild_failed

felixfontein commented:
ready_for_review

mattclay commented:
@felixfontein Can you explain what this feature would be used for?

felixfontein commented:
@mattclay the use-case is to use ansible-test to make sure that `verison_added` is added in collections for new options and modules with the corrrect value, and not modified for existing ones. Because right now, all `version_added` tests are restricted to ansible-base, so there are no checks at all for collections.

At the [last docs meeting](https://meetbot.fedoraproject.org/ansible-docs/2020-05-12/docs_working_group_aka_dawgs.2020-05-12-14.31.html), we agreed that `version_added` should be used with the collection version for new features.

felixfontein commented:
(Also, after the meeting @samccann mentioned that the ansible-maintained collection team also wants to use `version_added` this way.)

mattclay commented:
@felixfontein If I'm reading the changes correctly, it looks like this would enforce a workflow where the current collection version is expected to be the version of the next release. So collection authors would need to bump their collection version in their devel/master branch after a release. If they do not, then it would enforce the wrong version for version_added. Is that an accurate assessment?

felixfontein commented:
@mattclay that's why I added `--collection-version_added`, resp. made it a string option (and not boolean): the default value is `auto`, which will use the version from galaxy.yml and implies the workflow you describe.

But you can also explicitly specify a version which you expect as the value of `version_added` for new features/modules. That way, you do not need to update the version in `galaxy.yml` after release (or `galaxy.yml` can even not exist at all). Of course the value needs to come from somewhere, but all possible checks for correct values of `version_added` need to know it from somewhere. (Except if you're happy with "later thant the current version".)

Also, this is the reason why this check has to be enabled explicitly. If you use a workflow for a collection which doesn't store the value for the next release anywhere yet, you don't have to use this check.

(One last thing is that the version validation code is currently coupled with the version correctness code. The former should not be optional IMO and should always run. Changing that is somewhat more complex though, and I didn't manage to work on it during the week. I'll try to change that over the weekend. That way all `version_added` would always be checked to have correct semvers for collections, but the value for new options would only be compared to the wanted one if this is explicitly enabled.)

abadger commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/schema.py:
```diff
@@ -36,7 +36,28 @@ def semantic_version(v):
+    collection = m.group(1)
+    version = m.group(2)
+    if collection != 'ansible.builtin':
+        semantic_version(version)
```
Hmmm... if semantic_versions are followed, new features should always be inrroduced at  x.y.0, right?  I'm not sure if we want to enforce that, though.

(And I guess 0.y semantic versions are one wrinkle in that)

If we had a schema on docs parse, we could use it to allow "x.y" on parse but turn it into "x.y.0" before validation.  Unfortunately, that doesn't help with a schema that's only used on test.

felixfontein replied:
I don't mind if we enforce this. I guess deprecations should also only happen in x.y.0 versions (and removals in x.0.0 versions).

Maybe these should be optional errors though (i.e. only returned as errors when `--enable-optional-errors` is passed to ansible-test).

felixfontein commented:
This PR now (since yesterday, more precisely since 9c53fef1b251934c4ed6d1ec4ba946ae691be1da) tracks where `version_added` comes from. This is visible for example in ansible-doc: `ansible-doc community.crypto.openssl_privatekey` contains
```
- attributes
        The attributes the resulting file or directory should have. ...
        (Aliases: attr)[Default: (null)]
        type: str
        version_added: ansible.builtin:2.3
```
while version_added from the same collection is not prefixed:
```
- return_content
        If set ...
        [Default: False]
        type: bool
        version_added: 1.0.0
```
(I've chosen `ansible.builtin` for everything that comes from ansible-base.)

This is also used when validating `version_added` in ansible-test: `version_added` version numbers from collections has to conform to semantic versioning (i.e. `2.10` will error), while version numbers from ansible-base are not really validated. Also, the tests ensure that the top-level `version_added` must not come from a doc fragment of another collection (should be obvious, but currently that is possible).

felixfontein commented:
Results from discussion with @mattclay:
1. Split this PR up into two PRs:
    a) PR which traces version_added and does basic version_added validation (like making sure version numbers are of correct format);
    b) PR which allows to do version_added validation as in Ansible.
2. The PR from a) can make it into 2.10; the PR from b) will have to wait for 2.11.
3. For b), there will probably be a config files for collections (added before b) is completed), which will get a setting `version_added_workflow=[disabled|enabled|next_minor_version|next_major_version]` (default `disabled`). This is preferable over yet another CLI option.

TODO:
- split up PR.
- Ansible project meeting:
  - `version_added` ("magic") tracing? Docs team is in favor.
  - `removed_in_version` / `deprecated_aliases.version` explicit tracing? Also for `module.deprecate()`?

  See also https://github.com/ansible/community/wiki/Version-numbers-for-documentation-deprecation-removal-in-the-collection-world

felixfontein commented:
The first part of the split landed in #69631.

felixfontein commented:
This PR will stay in WIP mode until the other PR is done, then this one can be rebased / adjusted.

felixfontein commented:
#69680 (superset of #69631) has been merged, which does version_added tagging/tracing. Will rebase this PR next week.

mattclay commented:
We're going to hold on this until after 2.10. In 2.11 I'd like to add support for collections to have test configuration so they can define their workflow settings, and this is one of the PRs that would benefit from that.

felixfontein commented:
Depends on #70869, which fixes some assumptions made by this PR.

felixfontein commented:
/rebuild

felixfontein commented:
/rebuild_failed

felixfontein commented:
This PR currently has a problem with the symlinks in community.general; since we want to get rid of them soon anyway (for the next major release) I don't think it makes sense adding extra code to handle them.
